src/pycea/tl/permutation.py

```diff
@@ -0,0 +1,270 @@
+
+    Parameters
+    ----------
+    tdata
+        TreeData object.
+    keys
+        One or more `obs_keys`, `var_names`, `obsm_keys`, or `obsp_keys` to reconstruct.
+    reduction_fn
+        Callable function that can reduce the data from all the leaves of a given split to a vector or scalar. Defaults
+        to np.mean.
+    difference_fn
+        Callable function that takes to outputs from reduction_fn (one from each side of a node) and returns a scalar.
+        Defaults to left_stat - right_stat.
+    permutation_test
+        Whether to perform a permutation test to obtain a two-sided p-value for the
+        split statistic at each split node (subject to size constraints below).
+    n_permutations
+        Upper bound on the number of permutations to run. The actually executed
+        number is ``min(n_permutations, comb(n_left + n_right, n_left))`` per split.
+    min_required_permutations
+        Minimum number of **distinct** permutations required to run the test. If
+        ``comb(n_left + n_right, n_left) <= min_required_permutations``, the test
+        is skipped for that split.
+    keys_added
+        Attribute keys of `tdata.obst[tree].nodes` where split statistics will be stored. If `None`, `keys` are used.
+    tree
+        The `obst` key or keys of the trees to use. If `None`, all trees are used.
+    copy
+        If True, returns a :class:`DataFrame <pandas.DataFrame>` with split statistics.
+
+    Returns
+    -------
+    pd.DataFrame or None
+        If ``copy=True``, a DataFrame indexed by node identifiers with columns
+        ``f"{key_added}_split"`` and ``f"{key_added}_pval"`` for each requested key.
+        Nodes where the permutation test was skipped will have missing columns/values
+        for that key. If ``copy=False``, returns ``None`` after writing attributes
+        onto the graphs.
+    """
+
+    if isinstance(keys, str):
+        keys = [keys]
+    if keys_added is None:
+        keys_added = keys
+    if isinstance(keys_added, str):
+        keys_added = [keys_added]
+    if len(keys) != len(keys_added):
+        raise ValueError("Length of keys must match length of keys_added.")
+    if n_permutations < min_required_permutations:
+        raise ValueError("n_permutations must at least be min_required_permutations.")
+    tree_keys = tree
+    _check_tree_overlap(tdata, tree_keys)
+    trees = get_trees(tdata, tree_keys)
+
+    for key, key_added in zip(keys, keys_added):
+        data, is_array, is_square = get_keyed_obs_data(tdata, key)
+        if not(is_array or is_square):
+            data = data[key]
+        for _, t in trees.items():
+            for parent in nx.topological_sort(t):
+                children = list(t.successors(parent))
+
+                # don't do anything if not a split
+                if len(children) < 2:
+                    continue
+
+                # for each child, find all leaves
+                left_child = children[0]
```

Comment:
What if a node as >2 children? Currently the function just compares the first two without warning the user. A better approach maybe testing one-vs-rest for all the children. To maximize flexibility you could add a comparison argument
```
comparison: Literal["parent", "siblings"] = "siblings"
```
Where "siblings" is one-vs-rest and parent is comparison to the parent node